File: packages/pypi/agent_style/review/primitive.py
# SPDX-License-Identifier: MIT
"""High-level primitive entry points for audit and compare.

The primitive orchestrates:
  - loader.load_rules: find RULES.md (5-level resolution)
  - detectors_mech: regex / word-list detectors (deterministic)
  - detectors_struct: line/paragraph heuristics (deterministic)
  - detectors_sem: LLM-judge (stubbed; status "skipped" outside a skill host)

Both `audit` and `compare` emit the same per-rule violation schema so the CLI's
canonical-JSON contract is stable across invocations.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from typing import Optional

from agent_style.review.loader import LoadedRules, load_rules


@dataclass
class Violation:
    """One flagged instance of a rule in a file."""

    rule: str           # "RULE-01", "RULE-B", etc.
    line: int           # 1-based line number of the flagged content
    column: int         # 1-based column offset; 0 if line-level only
    excerpt: str        # the matched text, trimmed to ~120 chars
    detail: str         # short human note about why it matched


@dataclass
class RuleResult:
    """Per-rule audit result for one file."""

    rule: str
    severity: str
    detector: str               # "mechanical" | "structural" | "semantic"
    status: str                 # "ok" | "violation" | "skipped" | "error"
    count: int
    violations: list[Violation] = field(default_factory=list)
    note: str = ""              # populated for status "skipped" / "error"


@dataclass
class AuditResult:
    """Full audit of one file."""

    file: str
    rules_source: str
    rules_path: Optional[str]
    total_violations: int
    rule_results: list[RuleResult] = field(default_factory=list)


@dataclass
class CompareResult:
    """A/B comparison of two files."""

    file_a: str
    file_b: str
    rules_source: str
    rules_path: Optional[str]
    per_rule_delta: dict[str, dict[str, int]]   # {rule_id: {"a": n_a, "b": n_b, "delta": n_b - n_a}}
    total_a: int
    total_b: int


def audit(
    file_path: str,
    project_root: str = ".",
    mechanical_only: bool = False,
    audit_only: bool = True,
    skill_host: bool = False,
) -> AuditResult:
    """Run the deterministic (and optionally semantic) audit on one file.

    Outside a skill host (``skill_host=False``), semantic detectors are reported
    as ``status="skipped"``. Polish is never available from the primitive path;
    the CLI handles the --polish flag separately.

    If ``mechanical_only`` is True, structural detectors are also skipped
    (emitted as ``status="skipped"``). This mode is used by ``--cli-parity`` as
    the deterministic oracle.
    """
    rules = load_rules(project_root=project_root)
    text = _read_utf8(file_path)
    rule_results: list[RuleResult] = []
    total = 0

    # Lazy-import detectors to keep load_rules-only callers cheap.
    from agent_style.review import detectors_mech, detectors_struct, detectors_sem

    for rule in rules.rules:
        bucket = _classify(rule.id)
        if "mechanical" in bucket:
            result = detectors_mech.run(rule, text, file_path)
            rule_results.append(result)
            total += result.count
        if "structural" in bucket:
            if mechanical_only:
                rule_results.append(
                    RuleResult(
                        rule=rule.id,
                        severity=rule.severity,
                        detector="structural",
                        status="skipped",
                        count=0,
                        note="structural detectors disabled (--mechanical-only)",
                    )
                )
            else:
                result = detectors_struct.run(rule, text, file_path)
                rule_results.append(result)
                total += result.count
        if "semantic" in bucket:
            if mechanical_only or not skill_host:
                rule_results.append(
                    RuleResult(
                        rule=rule.id,
                        severity=rule.severity,
                        detector="semantic",
                        status="skipped",
                        count=0,
                        note=(
                            "semantic detectors require a skill host "
                            "(Claude Code or Anthropic Skills); "
                            "run /style-review via one of those hosts to include semantic rules"
                        ),
                    )
                )
            else:
                result = detectors_sem.run(rule, text, file_path)
                rule_results.append(result)
                total += result.count

    return AuditResult(
        file=file_path,
        rules_source=rules.rules_source,
        rules_path=rules.source_path,
        total_violations=total,
        rule_results=rule_results,
    )


def compare(
    file_a: str,
    file_b: str,
    project_root: str = ".",
    mechanical_only: bool = True,
    skill_host: bool = False,
) -> CompareResult:
    """Audit two files and compute per-rule violation deltas.

    Compare defaults to ``mechanical_only=True`` because benchmark / --cli-parity
    use cases need deterministic counts. Callers who want the semantic delta can
    set ``mechanical_only=False`` and ``skill_host=True``.
    """
    a = audit(file_a, project_root=project_root, mechanical_only=mechanical_only, skill_host=skill_host)
    b = audit(file_b, project_root=project_root, mechanical_only=mechanical_only, skill_host=skill_host)
    # Aggregate counts by rule ID across ALL buckets (mechanical + structural +
    # semantic). A rule that shows up in two buckets (e.g., RULE-05 has
    # mechanical + semantic) sums to a single per-rule count.
    per_rule: dict[str, dict[str, int]] = {}
    for ra in a.rule_results:
        row = per_rule.setdefault(ra.rule, {"a": 0, "b": 0, "delta": 0})
        row["a"] += ra.count
    for rb in b.rule_results:
        row = per_rule.setdefault(rb.rule, {"a": 0, "b": 0, "delta": 0})
        row["b"] += rb.count
    for row in per_rule.values():
        row["delta"] = row["b"] - row["a"]
    return CompareResult(
        file_a=file_a,
        file_b=file_b,
        rules_source=a.rules_source,
        rules_path=a.rules_path,
        per_rule_delta=per_rule,
        total_a=a.total_violations,
        total_b=b.total_violations,
    )


# ---------- Rule classification ---------------------------------------------
# Per Rev 3 detector matrix in PLAN-style-review.md. Each rule can have
# multiple buckets; both run for an audit and their counts aggregate.

_CLASSIFICATION: dict[str, set[str]] = {
    "RULE-01": {"semantic"},
    "RULE-02": {"structural", "semantic"},
    "RULE-03": {"semantic"},
    "RULE-04": {"semantic"},
    "RULE-05": {"mechanical", "semantic"},
    "RULE-06": {"mechanical", "semantic"},
    "RULE-07": {"structural"},
    "RULE-08": {"semantic"},
    "RULE-09": {"structural"},
    "RULE-10": {"structural"},
    "RULE-11": {"semantic"},
    "RULE-12": {"mechanical"},
    "RULE-A": {"structural"},
    "RULE-B": {"mechanical"},
    "RULE-C": {"structural"},
    "RULE-D": {"mechanical"},
    "RULE-E": {"structural"},
    "RULE-F": {"semantic"},
    "RULE-G": {"mechanical"},
    "RULE-H": {"semantic"},
    "RULE-I": {"mechanical"},
}


def _classify(rule_id: str) -> set[str]:
    return _CLASSIFICATION.get(rule_id, set())


def _read_utf8(path: str) -> str:
    with open(path, encoding="utf-8") as fh:
        return fh.read()

File: packages/npm/lib/review/primitive.js
// SPDX-License-Identifier: MIT
'use strict';
/**
 * High-level primitive for audit and compare.
 * Mirrors agent_style.review.primitive in the Python package.
 */

const fs = require('fs');
const { loadRules } = require('./loader');

// Per-rule bucket classification (same as Python).
const CLASSIFICATION = {
  'RULE-01': new Set(['semantic']),
  'RULE-02': new Set(['structural', 'semantic']),
  'RULE-03': new Set(['semantic']),
  'RULE-04': new Set(['semantic']),
  'RULE-05': new Set(['mechanical', 'semantic']),
  'RULE-06': new Set(['mechanical', 'semantic']),
  'RULE-07': new Set(['structural']),
  'RULE-08': new Set(['semantic']),
  'RULE-09': new Set(['structural']),
  'RULE-10': new Set(['structural']),
  'RULE-11': new Set(['semantic']),
  'RULE-12': new Set(['mechanical']),
  'RULE-A': new Set(['structural']),
  'RULE-B': new Set(['mechanical']),
  'RULE-C': new Set(['structural']),
  'RULE-D': new Set(['mechanical']),
  'RULE-E': new Set(['structural']),
  'RULE-F': new Set(['semantic']),
  'RULE-G': new Set(['mechanical']),
  'RULE-H': new Set(['semantic']),
  'RULE-I': new Set(['mechanical']),
};

function audit(filePath, opts = {}) {
  const projectRoot = opts.projectRoot || '.';
  const mechanicalOnly = opts.mechanicalOnly || false;
  const skillHost = opts.skillHost || false;

  const rules = loadRules({ projectRoot });
  const text = fs.readFileSync(filePath, 'utf8');
  const ruleResults = [];
  let total = 0;

  // Lazy-require detectors (avoids circular requires and keeps loader-only callers cheap)
  const detectorsMech = require('./detectors_mech');
  const detectorsStruct = require('./detectors_struct');
  const detectorsSem = require('./detectors_sem');

  for (const rule of rules.rules) {
    const bucket = CLASSIFICATION[rule.id] || new Set();
    if (bucket.has('mechanical')) {
      const result = detectorsMech.run(rule, text, filePath);
      ruleResults.push(result);
      total += result.count;
    }
    if (bucket.has('structural')) {
      if (mechanicalOnly) {
        ruleResults.push({
          rule: rule.id,
          severity: rule.severity,
          detector: 'structural',
          status: 'skipped',
          count: 0,
          note: 'structural detectors disabled (--mechanical-only)',
          violations: [],
        });
      } else {
        const result = detectorsStruct.run(rule, text, filePath);
        ruleResults.push(result);
        total += result.count;
      }
    }
    if (bucket.has('semantic')) {
      if (mechanicalOnly || !skillHost) {
        ruleResults.push({
          rule: rule.id,
          severity: rule.severity,
          detector: 'semantic',
          status: 'skipped',
          count: 0,
          note:
            'semantic detectors require a skill host (Claude Code or Anthropic Skills); ' +
            'run /style-review via one of those hosts to include semantic rules',
          violations: [],
        });
      } else {
        const result = detectorsSem.run(rule, text, filePath);
        ruleResults.push(result);
        total += result.count;
      }
    }
  }

  return {
    file: filePath,
    rulesSource: rules.rulesSource,
    rulesPath: rules.sourcePath,
    totalViolations: total,
    ruleResults,
  };
}

function compare(fileA, fileB, opts = {}) {
  const projectRoot = opts.projectRoot || '.';
  const mechanicalOnly = opts.mechanicalOnly !== undefined ? opts.mechanicalOnly : true;
  const skillHost = opts.skillHost || false;

  const a = audit(fileA, { projectRoot, mechanicalOnly, skillHost });
  const b = audit(fileB, { projectRoot, mechanicalOnly, skillHost });
  const perRule = {};

  for (const ra of a.ruleResults) {
    const row = perRule[ra.rule] || { a: 0, b: 0, delta: 0 };
    row.a += ra.count;
    perRule[ra.rule] = row;
  }
  for (const rb of b.ruleResults) {
    const row = perRule[rb.rule] || { a: 0, b: 0, delta: 0 };
    row.b += rb.count;
    perRule[rb.rule] = row;
  }
  for (const row of Object.values(perRule)) {
    row.delta = row.b - row.a;
  }

  return {
    fileA,
    fileB,
    rulesSource: a.rulesSource,
    rulesPath: a.rulesPath,
    totalA: a.totalViolations,
    totalB: b.totalViolations,
    perRuleDelta: perRule,
  };
}

module.exports = { audit, compare, CLASSIFICATION };
